File: Components/WinForms/CS/PivotGridOlapBrowser/Commands/BrowserCommandBarItems.cs
using DevExpress.Utils.Commands;
using DevExpress.XtraBars;
using DevExpress.XtraBars.Commands;
using DevExpress.XtraCharts;
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;

namespace PivotGridOlapBrowser.Commands.UI {
    #region IPivotGridBrowserBarItem interface
    public interface IPivotGridBrowserBarItem {
        PivotGridBrowser Browser { get; set; }
        BrowserCommandId CommandId { get; }
        void Update();
    }
    #endregion

    #region BrowserCommandBarButtonItem
    public abstract class BrowserCommandBarButtonItem : CommandBasedBarButtonItem, IPivotGridBrowserBarItem {
        BrowserCommandBarItemHelper browserHelper;

        public BrowserCommandBarButtonItem(BarManager manager, PivotGridBrowser pivotGridBrowser)
            : base(manager) {
            browserHelper = new BrowserCommandBarItemHelper(pivotGridBrowser);
            UpdateItemCaption();
        }

        protected PivotGridBrowser InternalBrowser {
            get { return browserHelper.Browser; }
            set {
                if(browserHelper.ChangeBrowser(value))
                    OnBrowserChanged();
            }
        }
        protected abstract BrowserCommandId InternalCommandId { get; }
        protected void Update() {
            UpdateItemVisibility();
        }
		protected virtual void OnBrowserChanged() {
		}
        protected override Command CreateCommand() {
            if(browserHelper == null) return null;
            return browserHelper.CreateCommand(InternalCommandId);
        }
        protected override void OnClick(BarItemLink link) {
            base.OnClick(link);
            ((BrowserCommandBarButtonItem)link.Item).InvokeCommand();
        }
        #region IPivotGridBrowserBarItem Members
        [Browsable(false), DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        PivotGridBrowser IPivotGridBrowserBarItem.Browser { get { return this.InternalBrowser; } set { this.InternalBrowser = value; } }

        BrowserCommandId IPivotGridBrowserBarItem.CommandId { get { return this.InternalCommandId; } }

        void IPivotGridBrowserBarItem.Update() { this.Update(); }
        #endregion

        #region IDisposable implementation
        protected override void Dispose(bool disposing) {
            try {
                if(disposing) {
                    browserHelper.ChangeBrowser(null);
                }
            } finally {
                base.Dispose(disposing);
            }
        }
        #endregion
    }
    #endregion

    #region BrowserCommandBarCheckItem
    public abstract class BrowserCommandBarCheckItem : CommandBasedBarCheckItem, IPivotGridBrowserBarItem {
        BrowserCommandBarItemHelper browserHelper;

        public BrowserCommandBarCheckItem(BarManager manager, PivotGridBrowser pivotGridBrowser)
            : base(manager) {
            browserHelper = new BrowserCommandBarItemHelper(pivotGridBrowser);
            UpdateItemCaption();
        }
                
        protected PivotGridBrowser InternalBrowser {
            get { return browserHelper.Browser; }
            set {
                if(browserHelper.ChangeBrowser(value))
                    OnBrowserChanged();
            }
        }
        protected abstract BrowserCommandId InternalCommandId { get; }
        protected abstract void ApplyChecking();

        protected void Update() {
            UpdateItemVisibility();
            UpdateItemChecked();
        }
        protected virtual void OnBrowserChanged() {
        }
        protected override void RaiseCheckedChanged() {
            base.RaiseCheckedChanged();
            ApplyChecking();
        }
        protected override Command CreateCommand() {
            if(browserHelper == null) return null;
            return browserHelper.CreateCommand(InternalCommandId);
        }

        #region IPivotGridBrowserBarItem Members
        [Browsable(false), DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        PivotGridBrowser IPivotGridBrowserBarItem.Browser { get { return this.InternalBrowser; } set { this.InternalBrowser = value; } }

        BrowserCommandId IPivotGridBrowserBarItem.CommandId { get { return this.InternalCommandId; } }

        void IPivotGridBrowserBarItem.Update() { this.Update(); }
        #endregion

        #region IDisposable implementation
        protected override void Dispose(bool disposing) {
            try {
                if(disposing) {
                    browserHelper.ChangeBrowser(null);
                }
            } finally {
                base.Dispose(disposing);
            }
        }
        #endregion
    }
    #endregion

    #region BrowserCommandBasedBarSubItem
    public abstract class BrowserCommandBasedBarSubItem : CommandBasedBarSubItem, IPivotGridBrowserBarItem {
        BrowserCommandBarItemHelper browserHelper;

        public BrowserCommandBasedBarSubItem(BarManager manager, PivotGridBrowser pivotGridBrowser)
            : base(manager) {
            browserHelper = new BrowserCommandBarItemHelper(pivotGridBrowser);
            UpdateItemCaption();
        }
                
        protected PivotGridBrowser InternalBrowser {
            get { return browserHelper.Browser; }
            set {
                if(browserHelper.ChangeBrowser(value))
                    OnBrowserChanged();
            }
        }
        protected abstract BrowserCommandId InternalCommandId { get; }

        protected void Update() {
            UpdateItemVisibility();
            UpdateItemChecked();
        }
        protected virtual void OnBrowserChanged() {
        }
        protected override Command CreateCommand() {
            if(browserHelper == null) return null;
            return browserHelper.CreateCommand(InternalCommandId);
        }

        #region IPivotGridBrowserBarItem Members
        [Browsable(false), DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        PivotGridBrowser IPivotGridBrowserBarItem.Browser { get { return this.InternalBrowser; } set { this.InternalBrowser = value; } }

        BrowserCommandId IPivotGridBrowserBarItem.CommandId { get { return this.InternalCommandId; } }

        void IPivotGridBrowserBarItem.Update() { this.Update(); }
        #endregion

        #region IDisposable implementation
        protected override void Dispose(bool disposing) {
            try {
                if(disposing) {
                    browserHelper.ChangeBrowser(null);
                }
            } finally {
                base.Dispose(disposing);
            }
        }
        #endregion
    }
    #endregion

    #region BrowserCommandBarItemHelper
    class BrowserCommandBarItemHelper {
        PivotGridBrowser browser;

        public BrowserCommandBarItemHelper(PivotGridBrowser browser) {
            this.browser = browser;
        }
        public Command CreateCommand(BrowserCommandId commandId) {
            if(browser == null) return null;
            return browser.CreateCommand(commandId);
        }
        public bool ChangeBrowser(PivotGridBrowser newBrowser) {
            if(Object.ReferenceEquals(this.browser, newBrowser))
                return false;
            UnsubscribeControlEvents();
            this.browser = newBrowser;
            if(this.browser != null)
                SubscribeControlEvents();
            return true;
        }
        public PivotGridBrowser Browser { get { return this.browser; } }

        protected virtual void SubscribeControlEvents() {
            if(this.browser == null) return;
            //TODO: browser.BeforeDispose += new EventHandler(...);
        }
        protected virtual void UnsubscribeControlEvents() {
            if(this.browser == null) return;
            //TODO: browser.BeforeDispose -= new EventHandler(...);
        }
    }
    #endregion

    #region ConnectToItem
    public class ConnectToItem : BrowserCommandBarButtonItem {
        public ConnectToItem(BarManager manager, PivotGridBrowser pivotGridBrowser)
            : base(manager, pivotGridBrowser) {
        }
        protected override BrowserCommandId InternalCommandId { get { return BrowserCommandId.ConnectTo; } }
    }
    #endregion

    #region CreateReportItem
    public class CreateReportItem : BrowserCommandBarButtonItem {
        public CreateReportItem(BarManager manager, PivotGridBrowser pivotGridBrowser)
            : base(manager, pivotGridBrowser) {
        }
        protected override BrowserCommandId InternalCommandId { get { return BrowserCommandId.CreateReport; } }
    }
    #endregion

    #region ShowChartFromPivotItem
    public class ShowChartFromPivotItem : BrowserCommandBarButtonItem {
        public ShowChartFromPivotItem(BarManager manager, PivotGridBrowser pivotGridBrowser)
            : base(manager, pivotGridBrowser) {
        }
        protected override BrowserCommandId InternalCommandId { get { return BrowserCommandId.ShowChartFromPivot; } }
    }
    #endregion
    
    #region ShowTotalsItem
    public class ShowTotalsItem : BrowserCommandBarCheckItem {
        public ShowTotalsItem(BarManager manager, PivotGridBrowser pivotGridBrowser)
            : base(manager, pivotGridBrowser) {
        }
        protected override BrowserCommandId InternalCommandId { get { return BrowserCommandId.ShowTotals; } }
        protected override void ApplyChecking() {
            InternalBrowser.ShowTotals = this.Checked;
        }
    }
    #endregion

    #region ShowSelectionItem
    public class ShowSelectionItem : BrowserCommandBarCheckItem {
        public ShowSelectionItem(BarManager manager, PivotGridBrowser pivotGridBrowser)
            : base(manager, pivotGridBrowser) {
        }
        protected override BrowserCommandId InternalCommandId { get { return BrowserCommandId.ShowSelection; } }
        protected override void ApplyChecking() {
            InternalBrowser.ShowSelection = this.Checked;
        }
    }
    #endregion

    #region ChartViewSetSubItem
    public class ChartViewSetSubItem : BrowserCommandBasedBarSubItem {
        public ChartViewSetSubItem(BarManager manager, PivotGridBrowser pivotGridBrowser)
            : base(manager, pivotGridBrowser) {
            InitSubItems(manager, pivotGridBrowser);
        }
        void InitSubItems(BarManager manager, PivotGridBrowser pivotGridBrowser) {
            Array viewTypes = Enum.GetValues(typeof(ViewType));
            foreach(object viewType1 in viewTypes) {
                ChartViewItem item = new ChartViewItem(manager, pivotGridBrowser, (ViewType)viewType1);
                AddItem(item);
                item.GroupIndex = 1;
                item.ItemClick += new ItemClickEventHandler(OnItemCheckedChanged);
                item.CheckedChanged += new ItemClickEventHandler(OnItemCheckedChanged);
                if(Comparer.Equals(item.ViewType, ViewType.Pie3D))
                    item.Checked = true;
            }
        }        
        protected void OnItemCheckedChanged(object sender, ItemClickEventArgs e) {
            ChartViewItem chartViewItem = (ChartViewItem)e.Item;
            Caption = GetDefaultCaption() + chartViewItem.ViewType.ToString();
        }
        protected override BrowserCommandId InternalCommandId { get { return BrowserCommandId.ChartViewSet; } }
    }
    #endregion
    #region ChartViewSubItemUIState
    public class ChartViewSubItemUIState : DevExpress.XtraBars.Commands.Internal.BarCheckItemUIState {//BarSubItemUIState
        public ChartViewSubItemUIState(ChartViewItem item)
            : base(item) {
        }
        public override object EditValue { 
            get {
                return Item.ViewType;
            } 
            set {
                Item.CheckViewType((ViewType)value);
            } 
        }
        protected new ChartViewItem Item { get { return (ChartViewItem)base.Item; } }
    }
    #endregion

    #region ChartViewItem
    public class ChartViewItem : BrowserCommandBarCheckItem {
        ViewType fViewType;

        public ChartViewItem(BarManager manager, PivotGridBrowser pivotGridBrowser, ViewType viewType)
            : base(manager, pivotGridBrowser) {
            this.fViewType = viewType;
            this.Caption = viewType.ToString();
        }
        public ViewType ViewType { get { return fViewType; } }
        public void CheckViewType(ViewType viewType) {
            if(Comparer.Equals(this.fViewType, viewType)) {
                this.Checked = true;
            }
        }
        protected override BrowserCommandId InternalCommandId { get { return BrowserCommandId.ChartView; } }
        protected override ICommandUIState CreateCommandUIState(Command command) {
            return new ChartViewSubItemUIState(this);
        }
        protected override void ApplyChecking() {
            InternalBrowser.ChartViewType = this.fViewType;
        }
    }
    #endregion
}

File: Components/WinForms/VB/PivotGridOlapBrowser/Commands/BrowserCommandBarItems.vb
Imports Microsoft.VisualBasic
Imports DevExpress.Utils.Commands
Imports DevExpress.XtraBars
Imports DevExpress.XtraBars.Commands
Imports DevExpress.XtraCharts
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.ComponentModel

Namespace PivotGridOlapBrowser.Commands.UI
	#Region "IPivotGridBrowserBarItem interface"
	Public Interface IPivotGridBrowserBarItem
		Property Browser() As PivotGridBrowser
		ReadOnly Property CommandId() As BrowserCommandId
		Sub Update()
	End Interface
	#End Region

	#Region "BrowserCommandBarButtonItem"
	Public MustInherit Class BrowserCommandBarButtonItem
		Inherits CommandBasedBarButtonItem
		Implements IPivotGridBrowserBarItem
		Private browserHelper As BrowserCommandBarItemHelper

		Public Sub New(ByVal manager As BarManager, ByVal pivotGridBrowser As PivotGridBrowser)
			MyBase.New(manager)
			browserHelper = New BrowserCommandBarItemHelper(pivotGridBrowser)
			UpdateItemCaption()
		End Sub

		Protected Property InternalBrowser() As PivotGridBrowser
			Get
				Return browserHelper.Browser
			End Get
			Set(ByVal value As PivotGridBrowser)
				If browserHelper.ChangeBrowser(value) Then
					OnBrowserChanged()
				End If
			End Set
		End Property
		Protected MustOverride ReadOnly Property InternalCommandId() As BrowserCommandId
		Protected Sub Update()
			UpdateItemVisibility()
		End Sub
		Protected Overridable Sub OnBrowserChanged()
		End Sub
		Protected Overrides Function CreateCommand() As Command
			If browserHelper Is Nothing Then
				Return Nothing
			End If
			Return browserHelper.CreateCommand(InternalCommandId)
		End Function
		Protected Overrides Sub OnClick(ByVal link As BarItemLink)
			MyBase.OnClick(link)
			CType(link.Item, BrowserCommandBarButtonItem).InvokeCommand()
		End Sub
		#Region "IPivotGridBrowserBarItem Members"
		<Browsable(False), DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)> _
		Private Property IPivotGridBrowserBarItem_Browser() As PivotGridBrowser Implements IPivotGridBrowserBarItem.Browser
			Get
				Return Me.InternalBrowser
			End Get
			Set(ByVal value As PivotGridBrowser)
				Me.InternalBrowser = value
			End Set
		End Property

		Private ReadOnly Property IPivotGridBrowserBarItem_CommandId() As BrowserCommandId Implements IPivotGridBrowserBarItem.CommandId
			Get
				Return Me.InternalCommandId
			End Get
		End Property

		Private Sub IPivotGridBrowserBarItem_Update() Implements IPivotGridBrowserBarItem.Update
			Me.Update()
		End Sub
		#End Region

		#Region "IDisposable implementation"
		Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
			Try
				If disposing Then
					browserHelper.ChangeBrowser(Nothing)
				End If
			Finally
				MyBase.Dispose(disposing)
			End Try
		End Sub
		#End Region
	End Class
	#End Region

	#Region "BrowserCommandBarCheckItem"
	Public MustInherit Class BrowserCommandBarCheckItem
		Inherits CommandBasedBarCheckItem
		Implements IPivotGridBrowserBarItem
		Private browserHelper As BrowserCommandBarItemHelper

		Public Sub New(ByVal manager As BarManager, ByVal pivotGridBrowser As PivotGridBrowser)
			MyBase.New(manager)
			browserHelper = New BrowserCommandBarItemHelper(pivotGridBrowser)
			UpdateItemCaption()
		End Sub

		Protected Property InternalBrowser() As PivotGridBrowser
			Get
				Return browserHelper.Browser
			End Get
			Set(ByVal value As PivotGridBrowser)
				If browserHelper.ChangeBrowser(value) Then
					OnBrowserChanged()
				End If
			End Set
		End Property
		Protected MustOverride ReadOnly Property InternalCommandId() As BrowserCommandId
		Protected MustOverride Sub ApplyChecking()

		Protected Sub Update()
			UpdateItemVisibility()
			UpdateItemChecked()
		End Sub
		Protected Overridable Sub OnBrowserChanged()
		End Sub
		Protected Overrides Sub RaiseCheckedChanged()
			MyBase.RaiseCheckedChanged()
			ApplyChecking()
		End Sub
		Protected Overrides Function CreateCommand() As Command
			If browserHelper Is Nothing Then
				Return Nothing
			End If
			Return browserHelper.CreateCommand(InternalCommandId)
		End Function

		#Region "IPivotGridBrowserBarItem Members"
		<Browsable(False), DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)> _
		Private Property IPivotGridBrowserBarItem_Browser() As PivotGridBrowser Implements IPivotGridBrowserBarItem.Browser
			Get
				Return Me.InternalBrowser
			End Get
			Set(ByVal value As PivotGridBrowser)
				Me.InternalBrowser = value
			End Set
		End Property

		Private ReadOnly Property IPivotGridBrowserBarItem_CommandId() As BrowserCommandId Implements IPivotGridBrowserBarItem.CommandId
			Get
				Return Me.InternalCommandId
			End Get
		End Property

		Private Sub IPivotGridBrowserBarItem_Update() Implements IPivotGridBrowserBarItem.Update
			Me.Update()
		End Sub
		#End Region

		#Region "IDisposable implementation"
		Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
			Try
				If disposing Then
					browserHelper.ChangeBrowser(Nothing)
				End If
			Finally
				MyBase.Dispose(disposing)
			End Try
		End Sub
		#End Region
	End Class
	#End Region

	#Region "BrowserCommandBasedBarSubItem"
	Public MustInherit Class BrowserCommandBasedBarSubItem
		Inherits CommandBasedBarSubItem
		Implements IPivotGridBrowserBarItem
		Private browserHelper As BrowserCommandBarItemHelper

		Public Sub New(ByVal manager As BarManager, ByVal pivotGridBrowser As PivotGridBrowser)
			MyBase.New(manager)
			browserHelper = New BrowserCommandBarItemHelper(pivotGridBrowser)
			UpdateItemCaption()
		End Sub

		Protected Property InternalBrowser() As PivotGridBrowser
			Get
				Return browserHelper.Browser
			End Get
			Set(ByVal value As PivotGridBrowser)
				If browserHelper.ChangeBrowser(value) Then
					OnBrowserChanged()
				End If
			End Set
		End Property
		Protected MustOverride ReadOnly Property InternalCommandId() As BrowserCommandId

		Protected Sub Update()
			UpdateItemVisibility()
			UpdateItemChecked()
		End Sub
		Protected Overridable Sub OnBrowserChanged()
		End Sub
		Protected Overrides Function CreateCommand() As Command
			If browserHelper Is Nothing Then
				Return Nothing
			End If
			Return browserHelper.CreateCommand(InternalCommandId)
		End Function

		#Region "IPivotGridBrowserBarItem Members"
		<Browsable(False), DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)> _
		Private Property IPivotGridBrowserBarItem_Browser() As PivotGridBrowser Implements IPivotGridBrowserBarItem.Browser
			Get
				Return Me.InternalBrowser
			End Get
			Set(ByVal value As PivotGridBrowser)
				Me.InternalBrowser = value
			End Set
		End Property

		Private ReadOnly Property IPivotGridBrowserBarItem_CommandId() As BrowserCommandId Implements IPivotGridBrowserBarItem.CommandId
			Get
				Return Me.InternalCommandId
			End Get
		End Property

		Private Sub IPivotGridBrowserBarItem_Update() Implements IPivotGridBrowserBarItem.Update
			Me.Update()
		End Sub
		#End Region

		#Region "IDisposable implementation"
		Protected Overrides Overloads Sub Dispose(ByVal disposing As Boolean)
			Try
				If disposing Then
					browserHelper.ChangeBrowser(Nothing)
				End If
			Finally
				MyBase.Dispose(disposing)
			End Try
		End Sub
		#End Region
	End Class
	#End Region

	#Region "BrowserCommandBarItemHelper"
	Friend Class BrowserCommandBarItemHelper
		Private browser_Renamed As PivotGridBrowser

		Public Sub New(ByVal browser As PivotGridBrowser)
			Me.browser_Renamed = browser
		End Sub
		Public Function CreateCommand(ByVal commandId As BrowserCommandId) As Command
			If browser_Renamed Is Nothing Then
				Return Nothing
			End If
			Return browser_Renamed.CreateCommand(commandId)
		End Function
		Public Function ChangeBrowser(ByVal newBrowser As PivotGridBrowser) As Boolean
			If Object.ReferenceEquals(Me.browser_Renamed, newBrowser) Then
				Return False
			End If
			UnsubscribeControlEvents()
			Me.browser_Renamed = newBrowser
			If Me.browser_Renamed IsNot Nothing Then
				SubscribeControlEvents()
			End If
			Return True
		End Function
		Public ReadOnly Property Browser() As PivotGridBrowser
			Get
				Return Me.browser_Renamed
			End Get
		End Property

		Protected Overridable Sub SubscribeControlEvents()
			If Me.browser_Renamed Is Nothing Then
				Return
			End If
			'TODO: browser.BeforeDispose += new EventHandler(...);
		End Sub
		Protected Overridable Sub UnsubscribeControlEvents()
			If Me.browser_Renamed Is Nothing Then
				Return
			End If
			'TODO: browser.BeforeDispose -= new EventHandler(...);
		End Sub
	End Class
	#End Region

	#Region "ConnectToItem"
	Public Class ConnectToItem
		Inherits BrowserCommandBarButtonItem
		Public Sub New(ByVal manager As BarManager, ByVal pivotGridBrowser As PivotGridBrowser)
			MyBase.New(manager, pivotGridBrowser)
		End Sub
		Protected Overrides ReadOnly Property InternalCommandId() As BrowserCommandId
			Get
				Return BrowserCommandId.ConnectTo
			End Get
		End Property
	End Class
	#End Region

	#Region "CreateReportItem"
	Public Class CreateReportItem
		Inherits BrowserCommandBarButtonItem
		Public Sub New(ByVal manager As BarManager, ByVal pivotGridBrowser As PivotGridBrowser)
			MyBase.New(manager, pivotGridBrowser)
		End Sub
		Protected Overrides ReadOnly Property InternalCommandId() As BrowserCommandId
			Get
				Return BrowserCommandId.CreateReport
			End Get
		End Property
	End Class
	#End Region

	#Region "ShowChartFromPivotItem"
	Public Class ShowChartFromPivotItem
		Inherits BrowserCommandBarButtonItem
		Public Sub New(ByVal manager As BarManager, ByVal pivotGridBrowser As PivotGridBrowser)
			MyBase.New(manager, pivotGridBrowser)
		End Sub
		Protected Overrides ReadOnly Property InternalCommandId() As BrowserCommandId
			Get
				Return BrowserCommandId.ShowChartFromPivot
			End Get
		End Property
	End Class
	#End Region

	#Region "ShowTotalsItem"
	Public Class ShowTotalsItem
		Inherits BrowserCommandBarCheckItem
		Public Sub New(ByVal manager As BarManager, ByVal pivotGridBrowser As PivotGridBrowser)
			MyBase.New(manager, pivotGridBrowser)
		End Sub
		Protected Overrides ReadOnly Property InternalCommandId() As BrowserCommandId
			Get
				Return BrowserCommandId.ShowTotals
			End Get
		End Property
		Protected Overrides Sub ApplyChecking()
			InternalBrowser.ShowTotals = Me.Checked
		End Sub
	End Class
	#End Region

	#Region "ShowSelectionItem"
	Public Class ShowSelectionItem
		Inherits BrowserCommandBarCheckItem
		Public Sub New(ByVal manager As BarManager, ByVal pivotGridBrowser As PivotGridBrowser)
			MyBase.New(manager, pivotGridBrowser)
		End Sub
		Protected Overrides ReadOnly Property InternalCommandId() As BrowserCommandId
			Get
				Return BrowserCommandId.ShowSelection
			End Get
		End Property
		Protected Overrides Sub ApplyChecking()
			InternalBrowser.ShowSelection = Me.Checked
		End Sub
	End Class
	#End Region

	#Region "ChartViewSetSubItem"
	Public Class ChartViewSetSubItem
		Inherits BrowserCommandBasedBarSubItem
		Public Sub New(ByVal manager As BarManager, ByVal pivotGridBrowser As PivotGridBrowser)
			MyBase.New(manager, pivotGridBrowser)
			InitSubItems(manager, pivotGridBrowser)
		End Sub
		Private Sub InitSubItems(ByVal manager As BarManager, ByVal pivotGridBrowser As PivotGridBrowser)
			Dim viewTypes As Array = System.Enum.GetValues(GetType(ViewType))
			For Each viewType1 As Object In viewTypes
				Dim item As New ChartViewItem(manager, pivotGridBrowser, CType(viewType1, ViewType))
				AddItem(item)
				item.GroupIndex = 1
				AddHandler item.ItemClick, AddressOf OnItemCheckedChanged
				AddHandler item.CheckedChanged, AddressOf OnItemCheckedChanged
				If Comparer.Equals(item.ViewType, ViewType.Pie3D) Then
					item.Checked = True
				End If
			Next viewType1
		End Sub
		Protected Sub OnItemCheckedChanged(ByVal sender As Object, ByVal e As ItemClickEventArgs)
			Dim chartViewItem As ChartViewItem = CType(e.Item, ChartViewItem)
			Caption = GetDefaultCaption() + chartViewItem.ViewType.ToString()
		End Sub
		Protected Overrides ReadOnly Property InternalCommandId() As BrowserCommandId
			Get
				Return BrowserCommandId.ChartViewSet
			End Get
		End Property
	End Class
	#End Region
	#Region "ChartViewSubItemUIState"
	Public Class ChartViewSubItemUIState 'BarSubItemUIState
		Inherits DevExpress.XtraBars.Commands.Internal.BarCheckItemUIState
		Public Sub New(ByVal item As ChartViewItem)
			MyBase.New(item)
		End Sub
		Public Overrides Property EditValue() As Object
			Get
				Return Item.ViewType
			End Get
			Set(ByVal value As Object)
				Item.CheckViewType(CType(value, ViewType))
			End Set
		End Property
		Protected Shadows ReadOnly Property Item() As ChartViewItem
			Get
				Return CType(MyBase.Item, ChartViewItem)
			End Get
		End Property
	End Class
	#End Region

	#Region "ChartViewItem"
	Public Class ChartViewItem
		Inherits BrowserCommandBarCheckItem
		Private fViewType As ViewType

		Public Sub New(ByVal manager As BarManager, ByVal pivotGridBrowser As PivotGridBrowser, ByVal viewType As ViewType)
			MyBase.New(manager, pivotGridBrowser)
			Me.fViewType = viewType
			Me.Caption = viewType.ToString()
		End Sub
		Public ReadOnly Property ViewType() As ViewType
			Get
				Return fViewType
			End Get
		End Property
		Public Sub CheckViewType(ByVal viewType As ViewType)
			If Comparer.Equals(Me.fViewType, viewType) Then
				Me.Checked = True
			End If
		End Sub
		Protected Overrides ReadOnly Property InternalCommandId() As BrowserCommandId
			Get
				Return BrowserCommandId.ChartView
			End Get
		End Property
		Protected Overrides Function CreateCommandUIState(ByVal command As Command) As ICommandUIState
			Return New ChartViewSubItemUIState(Me)
		End Function
		Protected Overrides Sub ApplyChecking()
			InternalBrowser.ChartViewType = Me.fViewType
		End Sub
	End Class
	#End Region
End Namespace
